Update README.md with DI content

lilyjma opened the pull request:
Adding dependency injection section
(The exact location to place the section is still to be determined - this is just an idea.)

# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

lilyjma committed:
Update README.md

adding dependency injection section

scottaddie commented on sdk/servicebus/Azure.Messaging.ServiceBus/README.md:
````diff
@@ -56,6 +56,36 @@ ServiceBusClient client = new ServiceBusClient(connectionString);
+  // code that uses ServiceBusClient
+}
+```
+For more details, see [Dependency injection with the Azure .NET SDK](https://docs.microsoft.com/dotnet/azure/sdk/dependency-injection).
````
```suggestion

For more details, see [Dependency injection with the Azure .NET SDK](https://docs.microsoft.com/dotnet/azure/sdk/dependency-injection).
```

scottaddie commented on sdk/servicebus/Azure.Messaging.ServiceBus/README.md:
```diff
@@ -56,6 +56,36 @@ ServiceBusClient client = new ServiceBusClient(connectionString);
+
+  public MyApiController(ServiceBusClient serviceBusClient)
+  {
+    _serviceBusClient = serviceBusClient;
```
```suggestion
      _serviceBusClient = serviceBusClient;
```

scottaddie reviewed (commented):
Nice work, Lily! Here are some suggestions to consider.

AlexGhiondea committed:
Update README.md

Proposing changes based on the chat with @KrzysztofCwalina and @davidfowl.

The changes are around simplifying the samples and using the configuration instead of the hard coded connection string.

scottaddie commented on sdk/servicebus/Azure.Messaging.ServiceBus/README.md:
```diff
@@ -56,6 +56,22 @@ ServiceBusClient client = new ServiceBusClient(connectionString);
+{
+    services.AddAzureClients(builder =>
+    {
+        builder.AddServiceBusClient(Configuration["connection_string_key"]);
```
@AlexGhiondea Should we use `Configuration.GetConnectionString("connection_string_key")` instead?

JoshLove-msft replied:
I think GetConnectionString assumes that the config includes a "ConnectionStrings" setting. We should probably include what the config looks like to go along with the code used.

scottaddie replied:
That's correct. We could use a config like this:

```json
{
  "ConnectionStrings": {
    "ServiceBus": "<connection_string>"
  }
}
```

Then the Service Bus client registration line becomes:

```csharp
builder.AddServiceBusClient(Configuration.GetConnectionString("ServiceBus"));
```

KrzysztofCwalina replied:
I agree showing the config json is a good idea. For some reason @davidfowl did not want to use .GetConnectionString and instead wanted the indexer. David, why? 
If a service supports AzureKeyCredential, the DI sample should use it instead of using the connection string. The type allows for rotating keys. I am fine with merging this as-is, but I would prefer if we filed and issue to update the sample later to use AzureKeyCredential.

KrzysztofCwalina replied:
@pakrym, @JoshLove-msft, I noticed the DI extension method for servicebus works only with connection strings and not with AzureKeyCredential. Do we not support rotating the keys at runtime?

JoshLove-msft replied:
The "rotating keys" part is on the AzureKeyCredential type. Are you envisioning that the user would create the AzureKeyCredential in their startup and then hold onto it? We could have an extension that creates the client using the KeyCredential, but it wouldn't be useful unless the users had access to the credential (which they wouldn't unless they created it themselves).

JoshLove-msft replied:
Service Bus does support AAD, so it might be worth showing that in addition to the connection string config.

KrzysztofCwalina replied:
> but it wouldn't be useful unless the users had access to the credential (which they wouldn't unless they created it themselves).

So, are you saying that if the user creates ServiceBusClient manually, they can rotate keys, but if they use DI they cannot? If that's the case, them it seems like a big hole in the DI story.

JoshLove-msft replied:
Yes, currently that is how it would work. In order to support KeyCredential DI I think it would have to either be injectable or retrievable from the Client. This is a general issue across any libraries using KeyCredential and not specific to Service Bus though.

KrzysztofCwalina replied:
I think we need a sample showing how to rotate keys with DI. Otherwise we would be potentially sending users onto a dead end road.

AlexGhiondea replied:
Updated the sample to account for some of this feedback. I think the sample for how to rotate should go into the larger document. Thoughts?

jsquire replied:
With respect to key rotation and such, it feels like we're getting too deep for the README if we start including that level of detail.   Perhaps we should just ensure that we have a sample dedicated to that concept and link to that from the README?

pakrym replied:
I agree with Jesse, we should have a more advanced sample of rolling AzureNamedKeyCredential but I don't think it should be in the main readme. 

Do we have examples of rolling `AzureNamedKeyCredential` without DI?

KrzysztofCwalina replied:
> With respect to key rotation and such, it feels like we're getting too deep for the README if we start including that level of detail. Perhaps we should just ensure that we have a sample dedicated to that concept and link to that from the README?


@jsquire, @pakrym, my point was that once we start pushing people toward DI, we need to explain how do do various things that are not very obvious when using DI. For example, changing a ctor-using sample to key rotation is something developers should be able to do on their own (if they understand basics of key rotation). Changing DI sample to key rotation is at least non-trivial if not outright impossible/impractical.

pakrym replied:
>my point was that once we start pushing people toward DI, we need to explain how do do various things that are not very obvious when using DI. 

Agreed.

AlexGhiondea committed:
Update README.md

Change the method named used and show the configuration snippet as well.

AlexGhiondea commented on sdk/servicebus/Azure.Messaging.ServiceBus/README.md:
````diff
@@ -56,6 +56,36 @@ ServiceBusClient client = new ServiceBusClient(connectionString);
+### Dependency Injection
+
+To inject ServiceBusClient as a dependency in an ASP .NET Core app, you can register the client in the `Startup.ConfigureServices` method in `Startup.cs`: 
+```C#
````
Fixed!

AlexGhiondea committed:
Update sdk/servicebus/Azure.Messaging.ServiceBus/README.md

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

AlexGhiondea committed:
Update sdk/servicebus/Azure.Messaging.ServiceBus/README.md

Co-authored-by: Scott Addie <10702007+scottaddie@users.noreply.github.com>

scottaddie committed:
minor edits

scottaddie reviewed (approved):
:shipit:

jsquire commented on sdk/servicebus/Azure.Messaging.ServiceBus/README.md:
````diff
@@ -56,6 +56,34 @@ ServiceBusClient client = new ServiceBusClient(connectionString);
+
+To inject `ServiceBusClient` as a dependency in an ASP.NET Core app, register the client in the `Startup.ConfigureServices` method:
+
+```csharp
````
I believe we'll want this content to be a snippet to ensure that we remain up-to-date and compliable.   I If we're doing this for multiple services,  I wonder if it makes sense to have an ASP.NET sample under the repository root `samples` directory and source all of the snippets from there?

pakrym replied:
I think it should be a part of each service's tests/samples as other snippets. We also need to devise a better way to test these extension methods.

lilyjma replied:
I agree with Pavel. Having this content in each service's readme will at least ensure that we don't miss the extension method for some packages (this was the case for Event Grid). Curious if different Azure SDK clients have different lifetime management needs. This might be another reason for each service to have its own content.

jsquire replied:
I wasn't suggesting that we move the README section itself,  just the code that we link to as a snippet.   This is currently inline, but we'll need to update so that it uses our snippet system so the code is compiled and validated during nightly runs.   

The question that I was posing was whether the test for that should live in the suite for each individual service or in some central location since there are dependencies for ASP.NET that would need to be included.

jsquire commented:
/azp run net - servicebus - ci
